content/blog/2022-05-19-CoC/index.md

```diff
@@ -0,0 +1,298 @@
+In CoC it is possible to define functions from terms to terms, types to types, types to terms, 
+and terms to types. Functions being able to go from terms to types is what makes CoC so powerful.
+Types that depend on terms are known as *dependent types*, and most languages
+do not have them. As an example, consider the following vector type in Coq
+where the length of the vector is part of its type.
+
+```coq
+Inductive vec : Type -> nat -> Type :=
+| vnil : forall X : Type, vec X 0
+| vcons : forall X : Type, X -> forall n : nat, vec X n -> vec X (S n).
+
+Definition v0 : vec bool 0 := vnil bool.
+Definition v1 : vec bool 1 := vcons bool true 0 v0.
+Fail Definition v0 : vec bool 1 := vnil bool.
+```
+
+Notice how the last definition fails because using the number 1 in `v0`'s type
+makes the term ill-typed.
+
+CoC is part of the Calculus of Inductive Constructions (CiC), which adds inductive
+types to CoC, and CiC is the basis of Gallina! But for the purposes of this
+project, we only implemented CoC.
+
+
+## Design and Implementation
+
+### Typing rules
+
+Unlike most other languages, CoC eliminates the distinction between types and terms. 
+We describe the CoC syntax below. We use $ M, N, ... $ for terms in general
+and $ x, y, z $ for variables.
+<p align="center">
+    $ M ::= x \; | \; (\lambda x:M) N \; | \; (M N) \; | \; [x:M] N \; | \; * $
+</p>
+
+$ [x:M] N $ is called a product, and $ * $ is the universe of all types, but is not a
+type itself. For an explanation of why, see the [Bonus section](#bonus).
+*Contexts* (denoted by $ \Gamma, \Delta $) are products over $*$ and have the
+form $[x_1:M_1] ... [x_n:M_n] *$. The inference system defines two judgements:
+
+1. $ \Gamma \vdash \Delta $ means $ \Delta $ is a valid context in the valid
+   context $ \Gamma $.
+2. $ \Gamma \vdash M : P $ means $ M $ is a well-typed term of type $ P $ in
+   the valid context $ \Gamma $
+
+In the inference system, a type $ P $ will either be a context or have the 
+property that $ \Gamma \vdash P : * $. Also note that $ [N/x]Q $ is the notation
+for substituting $N$ for $x$ in $Q$. The inference rules are below.
+
+<p align="center">
+<img src="fig1.png" alt="alt_text" title="image_tooltip" style="zoom:25%;" />
+</p>
+
+### Design
+
+When I started this project, I was writing my implementation based off of
+some notes written by Professor Clarkson, and this was the syntax of CoC that he used.
+
+<p align="center">
+    $ t := x \; | \; \lambda x:t_1 \rightarrow t_2 \; | \; (t_1 \; t_2) \; | \; \forall x:t_1, t_2 \; | \; \text{Type} $
+</p>
+
+I will be using this syntax throughout the rest of the blog post to go over examples,
+but the inference rules remain the same as before.
+
+I wanted the ability to write proofs in my programs, so I designed the AST with
+this in mind:
+```
+type t = 
+  | Id of id
+  | Fun of id * t * t
+  | App of t * t
+  | Forall of id * t * t
+  | Type
+
+and theorem =
+{
+  id: id;
+  theorem: t;
+  proof: t;
+}
```

Comment:
The concept of a "theorem" is a bit of a surprise here (although you mention it briefly above). Do you want to explain why this is in your AST, even though it's not in the CoC formalism you've described above? I know it resembles Coq, but being clear about how exactly this stuff maps on to the math you've just shown could help with readability. There is a an opportunity to make a bigger point here about what a "theorem" even _is_ in CoC-land…